### PR TITLE
Multithread analyzation and lazyload commits

### LINKED ISSUE #683: Big repos cause nodejs to run out of memory
Because all commits are kept in memory, we run out of memory for repos with 1 million+ commits.
This can be resolved by not keeping commits in memory for the entirety of hydration, but throwing them away after having analyzed their file changes.
This would mean that to show commits-history in git truck, commits must be loaded on demand

### PR DESCRIPTION
**Pull request by tjomson**
Resolves #683 

This pr adds major performance improvements due to multithreading of hydration.

The following table shows hydration time for tensorflow (156k commits) at different thread counts
<google-sheets-html-origin>
#of threads  | main (time sec) | multi-threaded (time sec)
-- | -- | --
1 | 177,61 | 176,43
2 | - | 134,65
4 | - | 76,4
8 | - | 59,79
16 | - | 55,18

Additionally, larger repos can now be analyzed due to lower memory usage, by loading commits on demand

### CODE REVIEW

File: src/analyzer/git-caller.server.ts [hunk 1]
```diff
@@ -264,6 +277,7 @@ export class GitCaller {
       `--max-count=${count}`,
       this.branch,
       "--numstat",
+      // "--cc", // include file changes for merge commits
```
Comment (joglr) [inline src/analyzer/git-caller.server.ts:280]: Please remove this commented-out code.

File: src/analyzer/git-caller.server.ts [hunk 2]
```diff
@@ -91,6 +91,19 @@ export class GitCaller {
...
+  async gitShow(commits: string[]) {
+    if (!this.branch) throw Error("branch not set")
+    const args = [
+      "show", 
+      "--no-patch", 
+      '--format="author <|%an|> date <|%at|> message <|%s|> body <|%b|> hash <|%H|>"'
+    ]
+
+    commits.forEach(x => args.push(x))
+    const result = (await runProcess(this.repo, "git", args)) as string
```
Comment (joglr) [inline src/analyzer/git-caller.server.ts:103]: Could the commits just be added to args using the spread operator?

File: src/analyzer/git-caller.server.ts [hunk 3]
```diff
@@ -356,8 +370,9 @@ export class GitCaller {
 
   async getCommitCount() {
     if (!this.branch) throw Error("Branch is undefined")
-    const result = await runProcess(this.repo, "git", ["rev-list", "--count", this.branch])
-    return result as number
+    const args = ["log", this.branch, '--format="%at"']
+    const result = await runProcess(this.repo, "git", args) as string
+    return result.split(/\s/).length
```
Comment (joglr) [inline src/analyzer/git-caller.server.ts:375]: Why was this changed? Is it faster than the old method?

File: src/analyzer/hydrate.server.ts [hunk 4]
```diff
@@ -1,100 +1,210 @@
...
       for (const contribMatch of contribMatches) {
         const file = contribMatch.groups?.file.trim()
         const isBinary = contribMatch.groups?.insertions === "-"
         if (!file) throw Error("file not found")
 
-        const hasBeenMoved = file.includes("=>")
 
         let filePath = file
-        if (hasBeenMoved) {
-          filePath = analyzeRenamedFile(filePath, renamedFiles)
```
Comment (joglr) [inline src/analyzer/hydrate.server.ts:58]: I think `hasBeenMoved` was more descriptive. Could we add a comment or add the variable back?

File: truckconfig.json [hunk 5]
```diff
@@ -11,8 +11,7 @@
     ]
   ],
   "hiddenFiles": [
```
Comment (joglr) [inline truckconfig.json:13]: This should be reverted, my bad.

File: src/analyzer/analyze.server.ts [hunk 6]
```diff
@@ -148,6 +149,7 @@ async function analyzeTree(path: string, name: string, hash: string) {
           name: newName,
           sizeInBytes: child.size as number,
           blameAuthors: {},
+          mutex: new Mutex()
```
Comment (joglr) [inline src/analyzer/analyze.server.ts:152]: Is the mutex necessary when there is no multithreading?

**Review by joglr — CHANGES_REQUESTED** [review]
Great job! Left a few questions

File: src/analyzer/git-caller.server.ts (hunk 1 above)
Comment (tjomson) [reply src/analyzer/git-caller.server.ts:280]: I would like to keep it in case we change our mind about including merge commits, or want to make it an option or something like that

File: src/analyzer/git-caller.server.ts (hunk 2 above)
Comment (tjomson) [reply src/analyzer/git-caller.server.ts:103]: indeed they could

File: src/analyzer/git-caller.server.ts (hunk 3 above)
Comment (tjomson) [reply src/analyzer/git-caller.server.ts:375]: I think at one point I wanted to be able to get the commit count of a file, which rev-list does not allow, but I no longer need to get the commit count for a file, so this should be reverted

File: src/analyzer/analyze.server.ts (hunk 6 above)
Comment (tjomson) [reply src/analyzer/analyze.server.ts:152]: Not sure, i guess it doesn't hurt. Haven't bothered to test O_O

Comment (tjomson) [reply src/analyzer/analyze.server.ts:152]: Just tested, it does not matter :(
it just seems way cooler with a mutex there

**Review by joglr — APPROVED** [review]
LGTM
